src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/DynamicStack.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Validation
+{
+    internal class DynamicStack
+    {
+        private const ushort CUTOFF = 20;
+
+        private List<object> _list;
+        private HashSet<object> _hashSet;
+        private bool _useHashSet = false;
+
+        public DynamicStack()
+        {
+            _list = new List<object>();
+        }
+
+        public int Count
+        {
+            get
+            {
+                if (_useHashSet)
+                {
+                    return _hashSet.Count;
+                }
+                else
+                {
+                    return _list.Count;
+                }
+            }
+        }
+
+        public void Push(object model)
+        {
+            if (_useHashSet)
+            {
+                _hashSet.Add(model);
+            }
+            else
+            {
+                _list.Add(model);
+            }
+
+            if (!_useHashSet && _list.Count > CUTOFF)
+            {
+                _hashSet = new HashSet<object>(_list);
```

Comment:
You need to use the reference equality comparer, we don't want to use overridden `Equals` methods